api/registrations/serializers.py

```diff
@@ -249,13 +251,36 @@ def create(self, validated_data):
         registration_choice = validated_data.pop('registration_choice', 'immediate')
         embargo_lifted = validated_data.pop('lift_embargo', None)
         reviewer = is_prereg_admin_not_project_admin(self.context['request'], draft)
+        children = validated_data.pop('children', None)
+        if children:
+            draft_root = draft.branched_from
+
+            # Validate to make sure we aren't registering nodes without registering their parents
+            # and all nodes have the same root.
+
+            # First check that all children are valid
+            child_nodes = Node.objects.filter(guids___id__in=children)
+            if child_nodes.count() != len(children):
+                raise exceptions.ValidationError('Some child nodes could not be found.')
+
+            # Second check that all children have a parent being registered. The exception being
+            #  the root node of the registration.
+            nodes_being_registered = list(child_nodes) + [draft_root]
+            for node in child_nodes:
+                parent_node = node.parent_node
+                if parent_node and parent_node not in nodes_being_registered and node != draft_root:
+                    raise exceptions.ValidationError('The parent of node {} must be registered.'.format(node._id))
+
+            excluded_node_ids = [node._id for node in draft_root.get_descendants_recursive() if node not in child_nodes]
```

Comment:
Same here: using a `set` will make the membership checks in `register_node` a bit faster.

Reply:
The flow here seems a bit convoluted.

1. Loading all the `Node`s whose _ids match `children`
2. Validate
3. Load the entire project hierarchy.
4. Get the _ids of the `Node`s in the project hierarchy that are not in `children`
5. (In register_node) Load the entire hierarchy again and register each node whose _id is not in the list from step 3.

There's a lot of redundant database work there. Would it make more sense to pass the "included" _ids to register?

```python
children = validated_data.pop('children', None)
...

try:
    draft.register(..., child_ids=children)
except InvalidChildError:
    raise ValidationError(....)
```

Then the flow would be:


1. Loading all the `Node`s whose _ids match `children`
2. (In register_node) Validate
3. (In register_node) Register each of the nodes from step 1.